infrastructure/merklemountainrange/src/merkle_change_tracker.rs

```diff
@@ -0,0 +1,232 @@
+    where
+        T: Serialize + DeserializeOwned,
+    {
+        if !self.enabled {
+            return Ok(());
+        }
+        self.current_horizon += 1;
+        let inc_index = self.current_horizon as i64 - self.pruning_horizon as i64;
+        if inc_index > 0 {
+            self.increase_pruning_horizon(inc_index as usize, hashmap, store)?
+        }
+
+        let mut checkpoint = MerkleCheckPoint {
+            objects_to_add: Vec::new(),
+            objects_to_del: Vec::new(),
+            mmr_to_add: Vec::new(),
+        };
+
+        checkpoint.objects_to_add.extend(self.objects_to_save.drain(..));
+        for hash in &checkpoint.objects_to_add {
+            let object = hashmap.get(hash);
+            if object.is_none() {
+                return Err(MerkleStorageError::SyncError);
+            }
+            let object = object.unwrap();
+            let _result = store.store(&to_hex(hash), &self.object_key, object)?;
+        }
+        checkpoint.objects_to_del.extend(self.objects_to_del.drain(..));
+        let mut counter = self.tree_saved;
+        while counter < mmr.len() {
+            checkpoint.mmr_to_add.push(mmr[counter].clone());
+            counter += 1;
+        }
+        let _result = store.store(&(self.current_horizon).to_string(), &self.mmr_key, &checkpoint)?;
+        let _result = store.store(&("init").to_string(), &self.init_key, &self.current_horizon)?;
+        self.tree_saved = counter - 1;
+
+        Ok(())
+    }
+
+    /// Function to load the checkpoint on pruning horizon and move that up
+    fn increase_pruning_horizon<T, S: MerkleStorage>(
+        &self,
+        index: usize,
+        hashmap: &mut HashMap<ObjectHash, MerkleObject<T>>,
+        store: &mut S,
+    ) -> Result<(), MerkleStorageError>
+    {
+        let mut cp = store.load::<MerkleCheckPoint>(&index.to_string(), &self.mmr_key)?;
+        let mut cp2 = store.load::<MerkleCheckPoint>(&(index + 1).to_string(), &self.mmr_key)?;
+        for hash in &cp2.objects_to_del {
+            let result = hashmap.remove(hash);
+            if result.is_none() {
+                return Err(MerkleStorageError::SyncError);
+            }
+        }
+        cp.add(&mut cp2);
+        let _result = store.store(&(index + 1).to_string(), &self.mmr_key, &cp)?;
+        let _result = store.delete(&(index).to_string(), &self.mmr_key)?;
+
+        Ok(())
+    }
+
+    /// Function to load an mmr
+    pub fn load<T, S: MerkleStorage>(
+        &mut self,
+        hashmap: &mut HashMap<ObjectHash, MerkleObject<T>>,
+        mmr: &mut Vec<MerkleNode>,
+        store: &mut S,
+    ) -> Result<(), MerkleStorageError>
+    where
+        T: Serialize + DeserializeOwned,
+    {
+        if !self.enabled {
+            return Ok(());
+        }
+        let amount_of_cps = store.load::<usize>(&("init").to_string(), &self.init_key)?;
+        self.current_horizon = if (amount_of_cps as i64 - self.pruning_horizon as i64) >= 0 {
```

Comment:
Idiomatic Rust:
```
self.current_horizon = match amount_of_cps.checked_sub(self.pruning_horizon) {
  None => 1,
  Some(v) => v,
}
```

Reply:
I have addressed this in a separate upcoming pr already